metadata_api.js

```diff
@@ -0,0 +1,80 @@
+const helpers = require('./helpers');
+const apiAdapter = require('./api_adapter');
+
+function getRawMetadata(token, bucket, name) {
+    return apiAdapter.getHeaders(
+        `https://${bucket}.storage.googleapis.com/${encodeURIComponent(name)}`,
+        `Bearer ${token}`
+    ).catch((e) => {
+        switch (e.status) {
+            case 403:
+                console.error('Permission denied for bucket object');
+                break;
+            default:
+                console.error('Unexpected error while trying to get headers for bucket object')
+        }
+        throw e;
+    });
+}
+
+function getPetTokenFromSam(bearerToken) {
+    return apiAdapter.postJsonTo(
+        `${helpers.samBaseUrl()}/api/google/v1/user/petServiceAccount/token`,
+        bearerToken,
+        '["https://www.googleapis.com/auth/devstorage.full_control"]')
+        .catch((e) => {
+            console.error('Failed to get Pet Service Account Token from Sam');
+            throw e;
+        });
+}
+
+function getGsObjectMetadata(gsUri, auth) {
+    const [bucket, name] = parseGsUri(gsUri);
+
+    return getPetTokenFromSam(auth)
+    .then((token) => getRawMetadata(token, bucket, name))
+    .then((response) => {
+        return {
+            contentType: response['content-type'],
+            size: parseInt(response['content-length']),
+            // timeCreated: ,
+            updated: response['last-modified'],
+            md5Hash: response['x-goog-hash'].substring(response['x-goog-hash'].indexOf('md5=') + 4),
+            bucket,
+            name,
+            uri: gsUri
+        };
+    })
+    .catch((e) => {
+        console.error(`Failed to get metadata for: ${gsUri}`);
+        throw e;
+    });
+}
+
+function parseGsUri(uri) {
+    return /gs:[/][/]([^/]+)[/](.+)/.exec(uri).slice(1);
+}
+
+function getGsUriFromDos(dosMetadata) {
+    return dosMetadata.urls.find((e) => e.url.startsWith('gs://')).url;
+}
+
+function getDosObject(dosUri) {
+    const newUri = helpers.dosToHttps(dosUri);
+
+    return apiAdapter.getJsonFrom(newUri).catch((e) => console.error('Failed while trying to retrieve DOS object', e));
+}
+
+async function getMetadata(url, auth, isDos) {
+    try {
+        const uri = isDos ?
+            getGsUriFromDos((await getDosObject(url)).data_object) :
```

Comment:
the auth passed in is not the correct auth to access the gs object returned by dos. Also the dos resolution contains most if not all the information required. Maybe this could be
`isDos ? getDosObjectMetadata : getGsObjectMetadata`

Reply:
The gs object itself is public, so we can use this auth to get its metadata, although not to get a signed url.

Reply:
no, it is not public... that's why we need to sign it. The one you are currently playing with may be public but the whole point of all this auth business is that they won't be.

Reply:
Ah, understood.